File: ModuleCommon/src/main/java/com/qihuan/commonmodule/base/BaseMvpFragment.java
package com.qihuan.commonmodule.base;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.view.View;

/**
 * BaseMvpFragment
 *
 * @author qi
 * @date 2018/5/28
 */
public abstract class BaseMvpFragment<P extends BasePresenter> extends BaseFragment implements BaseView {

    protected P mPresenter;

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        mPresenter = initPresenter();
        if (mPresenter != null) {
            // noinspection unchecked
            mPresenter.attachView(this);
        }
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        if (mPresenter != null) {
            mPresenter.detachView();
        }
    }

    @Override
    public void showLoading() {

    }

    @Override
    public void hideLoading() {

    }

    @Override
    public void showError(String errorMsg) {

    }

    protected abstract P initPresenter();

}

File: module_common/src/main/java/com/qihuan/commonmodule/base/BaseMvpFragment.kt
package com.qihuan.commonmodule.base

import android.os.Bundle
import android.view.View

/**
 * BaseMvpFragment
 *
 * @author qi
 * @date 2018/5/28
 */
abstract class BaseMvpFragment<in V : BaseView, out P : BasePresenter<V>> : BaseFragment(), BaseView {

    protected val mPresenter: P by lazy { initPresenter() }

    @Suppress("UNCHECKED_CAST")
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        mPresenter.attachView(this as V)
    }

    override fun onDestroyView() {
        super.onDestroyView()
        mPresenter.detachView()
    }

    override fun showLoading() {

    }

    override fun hideLoading() {

    }

    override fun showError(errorMsg: String) {

    }

    protected abstract fun initPresenter(): P

}
